### PR TITLE
BACKENDS: Add Win32 file browser and make system dialogs a backend feature

### PR DESCRIPTION
**Pull request by SupSuper**
I started with a simple dream: Using the Windows open dialog to add games to ScummVM instead of the GUI. And somehow ended up refactoring half the backend code. What happened?

Well, the current GUI `browser.cpp` supports either a ScummVM version or a OSX version. However it does so via platform `#ifdefs` which aren't very modular. I knew if I just added `WIN32` in there, it'd just be a big mess that nobody would want to support. So I tried to come up with a cleaner solution:

- Add a `DialogManager` to `OSystem` to handle calls to system dialogs (currently just the file browser), instead of polluting the GUI code. I tried to keep things consistent by following the example of existing managers like Update/Taskbar/etc.
- Move the existing OSX browser into a `MacOSXDialogManager`.
- Provide a `Win32DialogManager` implementation: https://i.imgur.com/M4VCiP6.png
I opted for the Vista+ common dialogs since they're more user-friendly than the older versions.

I also know people hate change, so I tried to keep this extra feature as optional and unobtrusive as possible:
- By default backends have a null `DialogManager`, and they can be omitted from code entirely via a `USE_DIALOGS` flag.
- If the backend can't provide a dialog, it falls back to the GUI version.
- If a user wants to use the GUI file browser regardless, they can turn off `gui_browser_native` (defaults to true).

Anyways this is my first time with ScummVM backend code so hopefully I didn't mess up too bad, looking forward to your feedback! I can only test this on Windows (MSVC and MinGW) so I'd appreciate feedback from other platforms.

### CODE REVIEW

**Comment by lotharsm** [comment]
Thanks @SupSuper for this great addition!

I gave it a quick test, everything seems to work great so far regarding the functionality.

However, one minor hiccup: When the ScummVM launcher is in fullscreen mode, opening the File Dialog causes the ScummVM launcher to return to windowed mode and then reverting back to fullscreen mode as soon as the dialog is closed again. Is this the desired behavior or something that could be adjusted?

**Comment by SupSuper** [comment]
@lotharsm The dialogs only work windowed. If ScummVM doesn't handle it, Windows will which is less pleasant. The behavior is consistent with the existing OSX file browser so it shouldn't be an issue.

I tried adding a checkbox but couldn't get it to fit properly in the Options, so I'll leave that to someone more experienced with the ScummVM GUI.

**Comment by Kirben** [comment]
Excellent addition for Windows, can finally directly access network drives, and special folders.
Short on time currently, so hopefully others can comment more code wise.

File: configure [hunk 1]
```diff
@@ -1003,6 +1004,7 @@ Optional Features:
   --disable-translation    don't build support for translated messages
   --disable-taskbar        don't build support for taskbar and launcher integration
   --disable-cloud          don't build cloud support
+  --disable-dialogs        don't build support for system dialogs
```
[the remark this answers is not in the document]
Comment (SupSuper) [reply configure]: Should I change the compiler macro as well? eg. `USE_SYSDIALOGS`

[the remark this answers is not in the document]
Comment (criezy) [reply configure]: I would say yes.

File: backends/dialogs/macosx/macosx-dialogs.mm [hunk 2]
```diff
@@ -175,25 +163,28 @@ - (IBAction) showHiddenFiles : (id) sender {
...
-	[panel setCanChooseFiles:!_isDirBrowser];
-	[panel setCanChooseDirectories:_isDirBrowser];
-	if (_isDirBrowser)
+	[panel setCanChooseFiles:!isDirBrowser];
+	[panel setCanChooseDirectories:isDirBrowser];
+	if (isDirBrowser)
 		[panel setTreatsFilePackagesAsDirectories:true];
-	[panel setTitle:(NSString *)_titleRef];
-	[panel setPrompt:(NSString *)_chooseRef];
+	[panel setTitle:titleRef];
```
Comment (criezy) [inline backends/dialogs/macosx/macosx-dialogs.mm]: This results in a compilation error. You need to keep the explicit cast to `NSString*` here (and on the next line).

**Comment by criezy** [comment]
Other than the compilation error I commented about, and the POTFILES change needed, it works well on macOS X. Once those two issues are fixed I will be happy with it.

I will let somebody else look and comment on the Windows code though.

Also it might be nice to have a command line parameter to turn this on and off, but this is something I can add at the same time as a GUI options once this is merged.

**Comment by SupSuper** [comment]
@criezy Thanks for testing, I've fixed the errors and cleaned up the commits to get rid of the unnecessary changes.

File: backends/dialogs/win32/win32-dialogs.cpp [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+		CoTaskMemFree(name);
+		delete[] str;
+	}
+	return hr;
+}
+
+Common::DialogManager::DialogResult Win32DialogManager::showFileBrowser(const char *title, Common::FSNode &choice, bool isDirBrowser) {
+	DialogResult result = kDialogError;
+
+	// Do nothing if not running on Windows Vista or later
```
Comment (bluegr) [inline backends/dialogs/win32/win32-dialogs.cpp]: Wouldn't it be better to just disable the feature altogether for earlier Windows versions? (i.e. XP or earlier)

**Comment by bluegr** [comment]
Nice work overall! I like the changes, and the overall design and code for this new feature.

Perhaps it would make sense to split commit a9fa7e581fcdffd732ceedb38344619177322a03 into a separate pull request

File: backends/dialogs/win32/win32-dialogs.cpp (hunk 3 above)
Comment (SupSuper) [reply backends/dialogs/win32/win32-dialogs.cpp]: I wanted to keep the option open for devs to add dialogs with different compatibility levels. Returning an error effectively disables the feature, as it'll fall back to the GUI dialog.

File: gui/browser.h [hunk 4]
```diff
@@ -37,35 +37,26 @@ class BrowserDialog : public Dialog {
...
 	Common::FSNode	_node;
-	Common::FSList			_nodeContent;
+	Common::FSList	_nodeContent;
+
 	bool _showHidden;
 	CheckboxWidget *_showHiddenWidget;
-#endif
+
 	Common::FSNode	_choice;
+	const char		*_title;
```
[the remark this answers is not in the document]
Comment (SupSuper) [reply gui/browser.h]: Fixed

File: backends/dialogs/win32/mingw-compat.h [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ *
+ */
+
+// TODO: Remove header when the latest changes to the Windows SDK have been integrated into MingW
+//       For reference, the interface definitions here are imported the SDK headers and from MingW-w64
```
[the remark this answers is not in the document]
Comment (SupSuper) [reply backends/dialogs/win32/mingw-compat.h]: Fixed

**Comment by SupSuper** [comment]
@bluegr a9fa7e5 is required for the Win32 Dialog Manager, so you would have to merge that before this one.

**Comment by digitall** [comment]
Since this has been open for several weeks and the issues seem to have been addressed, are there any objections to merging this now?

**Comment by criezy** [comment]
No objections from me.
I should also have some time this weekend to add a GUI options to enable or disable the use of the system file browser dialog. So if nobody merges this PR before, I will do it then.

**Comment by digitall** [comment]
@criezy: Will hold off until you add the GUI options, but think after that it should be fine to merge.

**Comment by criezy** [comment]
I will merge this now and work on the GUI option afterward to avoid a conflict on the theme zip files (those were changed since this PR was created).

**Comment by digitall** [comment]
@criezy : See https://bugs.scummvm.org/ticket/10875

**Pull request merged**